lib/web/app/match.go

```diff
@@ -63,6 +65,62 @@ func Match(ctx context.Context, authClient Getter, fn Matcher) ([]types.AppServe
 	return as, nil
 }
 
+// MatchRetryOptions retry options used on `MatchWithRetry` function.
+type MatchRetryOptions struct {
+	// Interval defines the interval between match retries.
+	Interval time.Duration
+	// MaxTime defines max time the match function will retry.
+	MaxTime time.Duration
+}
+
+// CheckAndSetDefaults checks and sets defaults.
+func (cfg *MatchRetryOptions) CheckAndSetDefaults() error {
+	if cfg.Interval == 0 {
+		return trace.BadParameter("must provide retry internal")
+	}
+	if cfg.MaxTime == 0 {
+		return trace.BadParameter("must provide retry max time")
+	}
+
+	return nil
+}
+
+// MatchWithRetry executes the `Match` function with a configurable retry
+// system.
+func MatchWithRetry(ctx context.Context, authClient Getter, fn Matcher, options MatchRetryOptions) ([]types.AppServer, error) {
+	ctx, cancel := context.WithTimeout(ctx, options.MaxTime)
+	defer cancel()
+
+	// Create a constant retry interval.
+	backoff, err := utils.NewConstant(options.Interval)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	var servers []types.AppServer
+	err = backoff.For(ctx, func() error {
```

Comment:
I'm not totally sure about this is approach. It has severals drawback: 
- If there are no any app service available the call will block the request fo -  10s 
- It works only for app access where this issue  occurs for db, windows desktop access.

Reply:
All solutions for retrying or just delaying the service startup are not reliable. I've done more extensive testing during the development of this PR. The time for a reverse tunnel to be established can vary depending on how the cluster is configured (more or less instances, for example). To give an example, even the default 10s couldn't prevent all errors from happening in some of my tests.

About your points:
* The retry is to give the service a room where the reverse tunnel could not be connected yet. However, that also means that if there isn't any app service up, all requests will exceed the timeout;
* Not sure if we can have something more generic, even if the implementation is placed on the service startup instead of per request. Still, each service has its startup logic (which can be shared, but we would have to update all of them);

That being said, I think we should work on some alternatives to this problem. Here is what I think would solve the issue: Change the startup logic of the service to wait until at least one reverse tunnel per registered application is up. It can take a long time if there are dynamic apps. But this way, we guarantee that the proxy is ready to serve any application request.

Reply:
> For example, even the default 10s couldn't prevent all errors from happening in some of my tests.

Wonder how this retry delay solution differs and  solves this case. Generally the `MatchWithRetry` function waits 10s till a reverse tunnel connection is established and app agent was successfully dialled.  I'm curious what make difference with postponing response from  service readiness probe vs current approach. Namely proxy discovery by an app agent take more than 10s current approach should also fail.

Reply:
For the most common scenario, both solutions would have similar effects but differences from the user's perspective. The rationale behind not adding the delay during the proxy start was that it could affect other proxy responsibilities (like authenticating users or serving DB access). In this way, we would isolate the delay just where needed. For example, if some proxy instances start to be killed for some reason, the newer ones would take longer to come online even if users are not using application access.

Reply:
> Not sure if I'm missing something but how an unstarted pod that doesn't handle any traffic can affect other proxy responsibilities? In my opinion adding an additional readyz probe delay affects only cluster scaling behavior.

The proxy won't be able to serve any other service (for example, database access).

But still, the delay solution won't cover all scenarios. Do you think we should skip them for now? or think about a more reliable way to define if the proxy is ready or not (like the one I mentioned in my previous comments)?